mathics/builtin/manipulate.py

```diff
@@ -0,0 +1,314 @@
+     = Manipulate[N[Sin[y]], {y, 1, 20, 2}]
+
+    >> Manipulate[i ^ 3, {i, {2, x ^ 4, a}}]
+     : Manipulate[] only works inside a Jupyter notebook.
+     = Manipulate[i ^ 3, {i, {2, x ^ 4, a}}]
+
+    >> Manipulate[x ^ y, {x, 1, 20}, {y, 1, 3}]
+     : Manipulate[] only works inside a Jupyter notebook.
+     = Manipulate[x ^ y, {x, 1, 20}, {y, 1, 3}]
+
+    >> Manipulate[N[1 / x], {{x, 1}, 0, 2}]
+     : Manipulate[] only works inside a Jupyter notebook.
+     = Manipulate[N[1 / x], {{x, 1}, 0, 2}]
+
+    >> Manipulate[N[1 / x], {{x, 1}, 0, 2, 0.1}]
+     : Manipulate[] only works inside a Jupyter notebook.
+     = Manipulate[N[1 / x], {{x, 1}, 0, 2, 0.1}]
+    """
+
+    # TODO: correct in the jupyter interface but can't be checked in tests
+    """
+    #> Manipulate[x, {x}]
+     = Manipulate[x, {x}]
+
+    #> Manipulate[x, {x, 1, 0}]
+     : 'Illegal variable range or step parameters for `x`.
+     = Manipulate[x, {x, 1, 0}]
+    """
+
+    attributes = ('HoldAll',)  # we'll call ReleaseHold at the time of evaluation below
+
+    messages = {
+        'jupyter': 'Manipulate[] only works inside a Jupyter notebook.',
+        'noipywidget': 'Manipulate[] needs the ipywidgets module to work.',
+        'widgetargs': 'Illegal variable range or step parameters for ``.'
+    }
+
+    def apply(self, expr, args, evaluation):
+        'Manipulate[expr_, args__]'
+
+        if (not _jupyter) or (not Kernel.initialized()) or (Kernel.instance() is None):
+            return evaluation.message('Manipulate', 'jupyter')
+
+        if not _ipywidgets:
+            return evaluation.message('Manipulate', 'noipywidget')
+
+        manip = Manipulations(evaluation)  # knows about the arguments and their widgets
+
+        for arg in args.get_sequence():
+            try:
+                if not manip.dispatch(arg.evaluate(evaluation), evaluation):  # not a valid argument pattern?
+                    return
+            except IllegalWidgetArguments as e:
+                evaluation.message('Manipulate', 'widgetargs', _strip_namespace(str(e.var)))
+
+        clear_output_callback = evaluation.clear_output_callback
+        display_data_callback = evaluation.display_data_callback  # for pushing updates
+
+        def callback(**kwargs):
+            clear_output_callback(wait=True)
+
+            line_no = evaluation.definitions.get_line_no()
+
+            new_evaluation = Evaluation(evaluation.definitions, result_callback=display_data_callback,
+                                        out_callback=evaluation.out_callback)
+
+            vars = [Expression('Set', Symbol(name), value) for name, value in kwargs.items()]
+            evaluatable = Expression('ReleaseHold', Expression('Module', Expression('List', *vars), expr))
+            new_evaluation.evaluate([evaluatable], timeout=settings.TIMEOUT)
+
+            evaluation.definitions.set_line_no(line_no)  # do not increment line_no for manipulate computations
```

Comment:
If we want to implement `Dynamic[]` or scheduled tasks, I can see us doing a lot of evaluation that isn't connected to a specific input line. Maybe we could make updating the line number a feature that can be disabled by something like `Evaluation(..., update_lineno=False)`?